pointmatcher/DataPointsFilters/Geometry.cpp

```diff
@@ -0,0 +1,170 @@
+    if (!cloud.descriptorExists("eigValues"))
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, no eigValues found in descriptors.");
+    }
+
+    // Validate descriptors and labels
+	int insertDim(0);
+	for(unsigned int i = 0; i < labelDim ; ++i)
+		insertDim += cloud.descriptorLabels[i].span;
+	if (insertDim != descDim)
+		throw InvalidField("SurfaceNormalDataPointsFilter: Error, descriptor labels do not match descriptor data");
+
+	// Reserve memory for new descriptors
+	const int dimSphericality(1);
+	const int dimUnstructureness(1);
+	const int dimStructureness(1);
+
+	boost::optional<View> sphericality;
+	boost::optional<View> unstructureness;
+	boost::optional<View> structureness;
+
+	Labels cloudLabels;
+	cloudLabels.push_back(Label("sphericality", dimSphericality));
+	if (keepUnstructureness)
+		cloudLabels.push_back(Label("unstructureness", dimUnstructureness));
+	if (keepStructureness)
+		cloudLabels.push_back(Label("structureness", dimStructureness));
+
+    // Reserve memory
+	cloud.allocateDescriptors(cloudLabels);
+
+	// Get the views
+    const View eigValues = cloud.getDescriptorViewByName("eigValues");
+    if (eigValues.rows() != 3)  // And check the dimensions
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, the number of eigValues is not 3.");
+    }
+
+	sphericality = cloud.getDescriptorViewByName("sphericality");
+	if (keepUnstructureness)
+		unstructureness = cloud.getDescriptorViewByName("unstructureness");
+	if (keepStructureness)
+		structureness = cloud.getDescriptorViewByName("structureness");
+
+    // Iterate through the point cloud and evaluate the geometry
+	for (int i = 0; i < pointsCount; ++i)
+	{
+	    // look at the three eigen values
+        Vector eig_vals_col = eigValues.col(i);
+        // might be already sorted but sort anyway
+        std::sort(eig_vals_col.data(),eig_vals_col.data()+eig_vals_col.size());
+
+        // finally, evaluate the geometry
+        T sphericality_val;
+        T unstructureness_val;
+        T structureness_val;
+
+        //TODO: Is there a more suitable limit for considering the values almost-zero? (VK)
+        if (fabs(eig_vals_col(2)) < std::numeric_limits<T>::min() or
+            fabs(eig_vals_col(1)) < std::numeric_limits<T>::min())
+        {
+            sphericality_val = std::numeric_limits<T>::quiet_NaN();
+            unstructureness_val = std::numeric_limits<T>::quiet_NaN();
+            structureness_val = std::numeric_limits<T>::quiet_NaN();
+        } else {
+            unstructureness_val = eig_vals_col(0) / eig_vals_col(2);
+            structureness_val =  (eig_vals_col(1) / eig_vals_col(2)) *
+                    ((eig_vals_col(1) - eig_vals_col(0)) / sqrt(eig_vals_col(0)*eig_vals_col(0) + eig_vals_col(1)*eig_vals_col(1)));
+            sphericality_val = unstructureness_val - structureness_val;
+        }
+
+        // store in the pointcloud
+        (*sphericality)(0,i) = sphericality_val;
```

Comment:
Is there a way to prevent this kind of access?

Reply:
I am afraid that there will always be some kind of dereferencing, since we are working with a view on the point cloud data... The star operator is overloaded from the boost optional thing to make it act as if dereferencing a pointer. I have changed the code to a method .get() call on the object, that look a little better (but still generates warnings for the next two optionals that they might not be initialized).